codersdk/workspaceagents.go

```diff
@@ -399,10 +414,23 @@ func runTailnetAPIConnector(
 		connected:     make(chan error, 1),
 		closed:        make(chan struct{}),
 	}
+	tac.gracefulCtx, tac.cancelGracefulCtx = context.WithCancel(context.Background())
+	go tac.manageGracefulTimeout()
 	go tac.run()
 	return tac
 }
 
+// manageGracefulTimeout allows the gracefulContext to last 1 second longer than the main context
+// to allow a graceful disconnect.
+func (tac *tailnetAPIConnector) manageGracefulTimeout() {
+	defer tac.cancelGracefulCtx()
+	<-tac.ctx.Done()
+	select {
+	case <-tac.closed:
+	case <-time.After(time.Second):
```

Comment:
(Non-blocking) I figure this is a best-effort situation, but will 1 second be enough? Does this need to be a configurable knob?

Reply:
I think it should be plenty, even on a slow connection because we're not waiting for a reply.  I definitely don't want to plumb configuration thru.

It is a best effort as you say --- consequence of not doing this is that the agent on the other side will see it as "lost" and possibly still try to handshake with it for up to 15 minutes.